addon/serializers/rest.js

```diff
@@ -200,8 +200,8 @@ var RESTSerializer = JSONSerializer.extend({
     const primaryHasTypeAttribute = modelHasAttributeOrRelationshipNamedType(primaryModelClass);
     // Support polymorphic records in async relationships
     if (!primaryHasTypeAttribute && hash.type && store._hasModelFor(this.modelNameFromPayloadKey(hash.type))) {
-      serializer = store.serializerFor(hash.type);
-      modelClass = store.modelFor(hash.type);
+      serializer = store.serializerFor(this.modelNameFromPayloadKey(hash.type));
+      modelClass = store.modelFor(this.modelNameFromPayloadKey(hash.type));
```

Comment:
I just realized this now, but at a first glance it looks weird the the modelName is normalized by a function called `modelNameFromPayloadKey`; since the passed argument is technically not a key from the payload.

That function originally has been added for the case where the payload key is different. So the `blog/post` entry in the following payload is used for the `post` model (see [this documentation](https://github.com/emberjs/data/blob/2322aacc3f3e4fb4fe9b30c7d790c45126fa5c73/addon/serializers/rest.js#L414-L471)):

```js
{
  "blog/post": {
    "id": 1
  }
}
```

`modelNameFromPayloadKey` is however already used within `json` serializer to normalize the modelName from `hash.type` (see [here](https://github.com/emberjs/data/blob/2322aacc3f3e4fb4fe9b30c7d790c45126fa5c73/addon/serializers/json.js#L603)); so this would not be the first time here that this function is used differently than the name would imply.

---

To be clear, I think it is definitely a bugfix to use the normalized modelName within the `if` branch here, so this PR looks good to me. I am just not sure if there should be a different function (like `modelNameFromPayloadType` or something), which would be more appropriate...